src/main/java/com/griddynamics/scenario/jagger/JHttpUserScenarioInvoker.java

```diff
@@ -26,29 +25,41 @@ public JHttpUserScenarioInvocationResult invoke(Void nothing, JHttpUserScenario
         Boolean scenarioSucceeded = true;
         JHttpUserScenarioStep previousStep = null;
         List<JHttpUserScenarioStepInvocationResult> stepInvocationResults = new ArrayList<>();
-        JHttpEndpoint globalEndpoint = scenario.getGlobalEndpoint();
 
-        for (JHttpUserScenarioStep userScenarioStep : scenario.userScenarioSteps) {
+        JHttpScenarioGlobalContext localScenarioContext = scenario.getScenarioGlobalContext();
+
+        for (JHttpUserScenarioStep userScenarioStep : scenario.getUserScenarioSteps()) {
             String metricId = getMetricId(scenario, userScenarioStep);
             String metricDisplayName = getMetricDisplayName(userScenarioStep);
 
             // Pre process step: internal setup. Can be later overridden by the user
             // Basic auth
-            if (scenario.getPassword() != null && scenario.getUserName() != null) {
-                String value = Base64.getEncoder().encodeToString((scenario.getUserName() + ":" + scenario.getPassword()).getBytes());
+            if (localScenarioContext.getPassword() != null && localScenarioContext.getUserName() != null) {
+                String value = Base64.getEncoder().encodeToString((localScenarioContext.getUserName() + ":" + localScenarioContext.getPassword()).getBytes());
                 userScenarioStep.getQuery().header("Authorization", "Basic " + value);
             }
 
-            // use global endpoint for step if present
-            if (globalEndpoint != null)
-                userScenarioStep.setEndpoint(globalEndpoint);
-            // check endpoint for null
-            if (globalEndpoint == null && userScenarioStep.getEndpoint() == null)
-                throw new IllegalArgumentException("Endpoint must not be null! Please, set global endpoint or set endpoint for every step.");
-
             // Pre process step: user actions executed before request
+            userScenarioStep.preProcessGlobalContext(previousStep, localScenarioContext);
```

Comment:
no. should it? this step is needed for updating context using previous step, as I understood.